README.md
# GstCurlHttpSrc

Fluendo's  GstCurlHttpSrc is a fork of [GstCurlHttpSrc](https://github.com/bbc/gst-curlhttpsrc)
It provides several features not found on the original plugin:

* The original plugin was downloading the whole content before sending downstream the single buffer.
  Now, as far as a new buffer arrives from cURL it is sent downstream
* The queue implementation was removed as it is no longer needed. In order to communicate
  between cURL and GStreamer we just use cond/mutex
* Support for GStreamer 0.10

## Here is the list properties the source element provides
* uri:
* username:
* password:
* proxyuri:
* proxyusername:
* proxypassword:
* cookies:
* useragent:
* headers:
* compress:
* redirect:
* maxredirect:
* keepalive:
* timeout:
* strict_ssl:
* ssl_ca_file:
* retries:
* connectionmaxtime:
* maxconcurrent_server:
* maxconcurrent_proxy:
* maxconcurrent_global:
* httpversion:


ChangeLog
2015-03-23  Josep Torra  <josep@fluendo.com>

	* m4/ipp.m4:
	Remove echo that wrongly slipped in with my previous commit.

2015-03-23  Josep Torra  <josep@fluendo.com>

	* m4/ipp.m4:
	Allow specify an specific CPU family with an optionar argument.

2015-03-21  Josep Torra  <josep@fluendo.com>

	* m4/ipp.m4:
	Prepend the library path.

2014-11-25  Andoni Morales <amorelas@fluendo.com>

	* m4/libtool.mk:
	Rename function with the same name

2014-11-25  Andoni Morales <amorelas@fluendo.com>

	* m4/libtool.mk:
	Append flags at the end of the command

2014-11-25  Andoni Morales <amorelas@fluendo.com>

	* m4/libtool.mk:
	parse inherited_linker_flags and include them in them
	in the link command

2014-11-24  Andoni Morales <amorelas@fluendo.com>

	* m4/libtool.mk:
	Add missing coma and include more debug

2014-11-24  Andoni Morales <amorelas@fluendo.com>

	* m4/libtool.mk:
	* m4/staticplugins.m4:
	Add support for tunning GStreamer libraries
	link type (static or dynamic)

2014-11-11  Josep Torra  <josep@fluendo.com>

	* gst-compat.h:
	Silence GST_TRACE when building with old GStreamer versions. 

2014-11-04  Josep Torra  <josep@fluendo.com>

	* gst-compat.h:
	Add missing backward compat definition.

2014-10-13  Xavi Artigas  <xartigas@fluendo.com>

	* gst-compat.h:
	Fix even more warnings (this one, and the two previous commits,
	are needed to build on Fedora5)

2014-10-13  Xavi Artigas  <xartigas@fluendo.com>

	* gst-demo.h:
	Fix warning

2014-10-13  Xavi Artigas  <xartigas@fluendo.com>

	* gst-compat.h:
	Fix warning

2014-10-03  Josep Torra  <josep@fluendo.com>

	* m4/gst-cpu-tune.m4:
	Drop -ftree-vectorize that slipped into common by error while
	experimenting with it.

2014-09-26  Josep Torra  <josep@fluendo.com>

	* gst-compat.h:
	Add gst_event_has_name replacement for older GStreamer 0.10 versions.

2014-07-21  Xavi Artigas  <xartigas@fluendo.com>

	* gst-demo.h:
	Fix precedence of = and ==. This only triggered a problem when
	compiling the h265 plugin, because it uses C++ and the result of
	a == is a boolean, not an int which can be silently converted
	to an enum.

2014-06-19  Josep Torra  <josep@fluendo.com>

	* gst-compat.h:
	Provide replacement for gst_adapter_unmap in GStreamer 0.10.

2014-05-14  Andoni Morales  <amorales@fluendo.com>

	* m4/staticplugins.m4:
	Add replacement for d3dvideosink

2014-05-12  Andoni Morales  <amorales@fluendo.com>

	* gst-compat.h:
	Fix compiler error casting from gpointer to guint8 *

2014-05-07  Josep Torra  <josep@fluendo.com>

	* gst-compat.h:
	Add few more replacements for 0.10.

2014-05-07  Josep Torra  <josep@fluendo.com>

	* gst-compat.h:
	Add gst_buffer_new_wrapped replacement for 0.10.

2014-04-02  Jorge Zapata  <jzapata@fluendo.com>

	* m4/ipp.m4:
	Do not use ipp on non-intel archs

2014-04-01  Josep Torra  <josep@fluendo.com>

	* gst-compat.h:
	Add a replacement for g_list_free_full in order to allow builds
	with older glib2.

2014-04-01  Josep Torra  <josep@fluendo.com>

	* gst-compat.h:
	Provide guintptr definition to allow builds wiht older glib2.

2014-03-25  Josep Torra  <josep@fluendo.com>

	* gst-compat.h:
	Add a replacement for gst_caps_steal_structure.

2014-03-14  Jorge Zapata <jzapata@fluendo.com>

	* doc:
	* doc/Doxyfile_common.in:
	* doc/css:
	* doc/css/stylesheet.css:
	* doc/doxygen_to_devhelp.xsl:
	* doc/footer.html:
	* doc/header.html:
	* doxygen.mak:
	Add the documentation common files

2014-01-28  Andoni Morales <amorales@fluendo.com>

	* m4/platform.m4: Add detection support for iOS 

2014-01-18  Julien Moutte <julien@fluendo.com>

	* m4/ipp.m4: when linking on Windows we need to make
	sure that we pull symbols from MSVCRT instead of NTDLL.
	Indeed I've seen sscanf from NTDLL seriously break and the
	symbol was used from NTDLL because of our link flag from ipp.

2014-01-16  Andoni Morales  <amorales@fluendo.com>

	* m4/platform.m4:
	Add solaris platform detection

2013-12-27  Andoni Morales  <amorales@fluendo.com>

	* wibu.m4:
	Add framework search path for Mavericks

2013-12-27  Andoni Morales  <amorales@fluendo.com>

	* wibu.m4:
	Fix library names in Windows

2013-12-27  Xavi Artigas  <xartigas@fluendo.com>

	* m4/wibu.m4:
	Use correct WIBU CodeMeter Linux library.

2013-12-24  Andoni Morales  <amorales@fluendo.com>

	* wibu.m4:
	Add an option to check for AxProtector, CodeMeter or both

2013-12-24  Andoni Morales  <amorales@fluendo.com>

	* platform.m4:
	Add macro for platform detection
	* wibu.m4:
	Add macro for WIBU detection

2013-12-20  Josep Torra  <josep@fluendo.com>

	* gst-compat.h:
	Add gst_caps_set_value replacement for older GStreamer versions.

2013-11-29  Julien Moutte <julien@fluendo.com>

	* fluendo-plugins-sdk.props: Prepare for 1.0 build while maintaining
	backward compatibility with old project files.

2013-11-25  Andoni Morales <amorales@fluendo.com>

	* m4/staticplugins.am:
	Fix registry names for windows plugins

2013-11-22  Andoni Morales <amorales@fluendo.com>

	* m4/libtool.mk: Remove '=' from search paths (as seen in windows)

2013-11-15  Andoni Morales <amorales@fluendo.com>

	* m4/staticplugins.am: Use the more portable printf and
	fix plugin regsitry entry point for osxvideosink

2013-11-15  Andoni Morales <amorales@fluendo.com>

	* m4/libtool.mk: Add missing -l to shared libraries that don't have it

2013-11-15  Andoni Morales <amorales@fluendo.com>

	* m4/libtool.mk: Import helper from the android ndk
	* m4/staticplugins.am:
	Add a new helper to transform a list of plugins into
	a linker command linking gstreamer and its plugins
	statically and the rest as shared libraries

2013-11-15  Andoni Morales <amorales@fluendo.com>

	* m4/staticplugins.am: Automake helper for static plugins

2013-11-04  Julien Moutte <julien@fluendo.com>

	* gst-demo.h: Need to include the base class for audio encoder.

2013-11-04  Julien Moutte <julien@fluendo.com>

	* gst-demo.h: Fix small mistakes in my previous commit for 1.0.

2013-11-04  Julien Moutte <julien@fluendo.com>


gst-compat.h
/*
 * FLUENDO S.A.
 * Copyright (C) <2012>  <support@fluendo.com>
 */

#ifndef GST_COMPAT_H
#define GST_COMPAT_H

#include <string.h>

#include <gst/gst.h>
#include <gst/base/gstadapter.h>

#if !GLIB_CHECK_VERSION(2,14,0)
#define g_once_init_enter(_token) (!(*(_token)))
#define g_once_init_leave(_token,_value) (*(_token)=(_value))
#endif

#if !GLIB_CHECK_VERSION(2,18,0)
typedef unsigned long guintptr;
#endif

#if (!GLIB_CHECK_VERSION(2,28,0))
static inline void
g_list_free_full (GList * list, GDestroyNotify free_func)
{
  GList *next = list;
  while (next) {
    free_func (next->data);
    next = g_list_remove_link (next, next);
  }
}
#endif

#ifndef GST_CHECK_VERSION
#define GST_CHECK_VERSION(major,minor,micro)  \
    (GST_VERSION_MAJOR > (major) || \
     (GST_VERSION_MAJOR == (major) && GST_VERSION_MINOR > (minor)) || \
     (GST_VERSION_MAJOR == (major) && GST_VERSION_MINOR == (minor) && \
      GST_VERSION_MICRO >= (micro)))
#endif

#if !GST_CHECK_VERSION(0,10,4)
#define GST_QUERY_TYPE_NAME(query) (gst_query_type_get_name(GST_QUERY_TYPE(query)))
#endif

#if !GST_CHECK_VERSION(0,10,6)
static inline GstBuffer *
gst_adapter_take_buffer (GstAdapter * adapter, guint nbytes)
{
  GstBuffer *buf = NULL;

  if (G_UNLIKELY (nbytes > adapter->size))
    return NULL;

  buf = gst_buffer_new_and_alloc (nbytes);

  if (G_UNLIKELY (!buf))
    return NULL;

  /* Slow... */
  memcpy (GST_BUFFER_DATA (buf), gst_adapter_peek (adapter, nbytes), nbytes);

  return buf;
}
#endif

#if !GST_CHECK_VERSION(0,10,7)
#define GST_FLOW_CUSTOM_SUCCESS ((GstFlowReturn)100)
#define GST_FLOW_CUSTOM_ERROR ((GstFlowReturn)-100)
#define GST_FLOW_IS_SUCCESS(ret) ((ret) >= GST_FLOW_OK)
#endif

#if !GST_CHECK_VERSION(0,10,9)
#define GST_BUFFER_IS_DISCONT(buffer) \
    (GST_BUFFER_FLAG_IS_SET (buffer, GST_BUFFER_FLAG_DISCONT))
#endif

#if !GST_CHECK_VERSION(0,10,10)
static inline GstPad *
gst_ghost_pad_new_from_template (const gchar * name, GstPad * target,
    GstPadTemplate * templ)
{
  GstPad *ret;

  g_return_val_if_fail (GST_IS_PAD (target), NULL);
  g_return_val_if_fail (!gst_pad_is_linked (target), NULL);
  g_return_val_if_fail (templ != NULL, NULL);
  g_return_val_if_fail (GST_PAD_TEMPLATE_DIRECTION (templ) ==
      GST_PAD_DIRECTION (target), NULL);

  if ((ret = gst_ghost_pad_new_no_target (name, GST_PAD_DIRECTION (target)))) {
    if (!gst_ghost_pad_set_target (GST_GHOST_PAD (ret), target))
      goto set_target_failed;
    g_object_set (ret, "template", templ, NULL);
  }

  return ret;

  /* ERRORS */
set_target_failed:
  {
    gst_object_unref (ret);
    return NULL;
  }
}
#endif

#if !GST_CHECK_VERSION(0,10,11)
#define gst_message_new_buffering(elem,perc) \
    gst_message_new_custom (GST_MESSAGE_BUFFERING,         \
        (elem),                                            \
        gst_structure_new ("GstMessageBuffering",          \
            "buffer-percent", G_TYPE_INT, (perc), NULL))

#endif

#if !GST_CHECK_VERSION(0,10,14)
#define gst_element_class_set_details_simple(klass,longname,classification,description,author) \
    G_STMT_START{                                                             \
      static GstElementDetails details =                                      \
          GST_ELEMENT_DETAILS (longname,classification,description,author);   \
      gst_element_class_set_details (klass, &details);                        \
    }G_STMT_END
#endif

#if !GST_CHECK_VERSION(0,10,15)
#define gst_structure_get_uint(stru,fn,fv) \
    gst_structure_get_int(stru,fn,(gint*)fv)
#endif

#if !GST_CHECK_VERSION(0,10,20)
static inline gboolean
gst_event_has_name (GstEvent * event, const gchar * name)
{
  g_return_val_if_fail (GST_IS_EVENT (event), FALSE);

  if (event->structure == NULL)
    return FALSE;

  return gst_structure_has_name (event->structure, name);
}
#endif

#if !GST_CHECK_VERSION(0,10,23)
#define GST_BUFFER_FLAG_MEDIA1 (GST_MINI_OBJECT_FLAG_LAST << 5)
#define GST_BUFFER_FLAG_MEDIA2 (GST_MINI_OBJECT_FLAG_LAST << 6)
#define GST_BUFFER_FLAG_MEDIA3 (GST_MINI_OBJECT_FLAG_LAST << 7)

#define GST_MEMDUMP(_title, _data, _length) while (0)
#define GST_MEMDUMP_OBJECT(_object, _title, _data, _length) while (0)
#endif

#if !GST_CHECK_VERSION(0,10,24)
static inline void
gst_object_ref_sink (gpointer object)
{
  g_return_if_fail (GST_IS_OBJECT (object));

  GST_OBJECT_LOCK (object);
  if (G_LIKELY (GST_OBJECT_IS_FLOATING (object))) {
    GST_OBJECT_FLAG_UNSET (object, GST_OBJECT_FLOATING);
    GST_OBJECT_UNLOCK (object);
  } else {
    GST_OBJECT_UNLOCK (object);
    gst_object_ref (object);
  }
}
#endif

#if !GST_CHECK_VERSION(0,10,26)
static inline void
gst_caps_set_value (GstCaps *caps, const char *field, const GValue *value)
{
  GstStructure *structure;
  structure = gst_caps_get_structure (caps, 0);
  gst_structure_set_value (structure, field, value);
}
#endif

#if !GST_CHECK_VERSION(0,10,30)
static inline GstStructure *
gst_caps_steal_structure (GstCaps * caps, guint index)
{
  GstStructure *s;
  g_return_val_if_fail (caps != NULL, NULL);
  g_return_val_if_fail ((g_atomic_int_get (&(caps)->refcount) == 1), NULL);

  if (G_UNLIKELY (index >= caps->structs->len))
    return NULL;

  s = (GstStructure *)g_ptr_array_remove_index (caps->structs, index);
  gst_structure_set_parent_refcount (s, NULL);
  return s;
}
#endif

#if !GST_CHECK_VERSION(0,10,33)
#define GST_MINI_OBJECT_FLAG_RESERVED1 (1<<1)
#define GST_BUFFER_FLAG_MEDIA4 GST_MINI_OBJECT_FLAG_RESERVED1
#endif

#if !GST_CHECK_VERSION (1,0,0)

typedef struct {
  guint8 * data;
  gsize size;
} GstMapInfo;

typedef enum {
  GST_MAP_READ = 1 << 0,
  GST_MAP_WRITE = 1 << 1
} GstMapFlags;

static inline gboolean
gst_buffer_map (GstBuffer * buffer, GstMapInfo * info, GstMapFlags flags)
{
  info->data = GST_BUFFER_DATA (buffer);
  info->size = GST_BUFFER_SIZE (buffer);
  return TRUE;
}

#define gst_buffer_unmap(buffer,info) while(0)
#define gst_buffer_get_size(buffer) ((gsize) GST_BUFFER_SIZE((buffer)))

#define GST_FLOW_EOS GST_FLOW_UNEXPECTED
#define GST_FLOW_FLUSHING GST_FLOW_WRONG_STATE

static inline GstBuffer *
gst_buffer_new_wrapped (gpointer data, gsize size)
{
  GstBuffer * buffer = gst_buffer_new ();

  GST_BUFFER_DATA (buffer) = GST_BUFFER_MALLOCDATA (buffer) = (guint8 *) data;
  GST_BUFFER_SIZE (buffer) = size;

  return buffer;
}

#define gst_adapter_map gst_adapter_peek
#define gst_adapter_unmap(adapter) while(0)
#define gst_segment_do_seek gst_segment_set_seek
#define gst_tag_list_new_empty gst_tag_list_new

#endif
#endif /* GST_COMPAT_H */
